src/storage/src/table/mod.rs

```diff
@@ -108,12 +112,19 @@ pub trait TableIter: Send {
 
 /// Get vnode value with `indices` on the given `row`.
 pub fn compute_vnode(row: impl Row, indices: &[usize], vnodes: &Bitmap) -> VirtualNode {
-    let vnode = if indices.is_empty() {
-        DEFAULT_VNODE
-    } else {
-        let vnode = (&row).project(indices).hash(Crc32FastBuilder).to_vnode();
+    let vnode = if !indices.is_empty() {
+        let project = (&row).project(indices);
+
+        let vnode = if let Ok(d) = project.iter().exactly_one() && let Some(ScalarRefImpl::Serial(s)) = d.as_ref() {
+            HashCode::from(extract_vnode_id_from_row_id(s.as_row_id()) as u64).to_vnode()
+        } else {
+            project.hash(Crc32FastBuilder).to_vnode()
```

Comment:
Here I prefer to abandon CRC32 totally.

For batch query system, it doesn't matter because it's only a temporary stuff. But for RisingWave, persisted data (tables & MVs) is distributed by `Hash(PK)`, it had better to be more understandable and predictable.

For example, if you define a table that `partitioned by hash(id)` in MySQL/PG, you would expect `id: 0` in partition 0, `id: 1` in partition 1, and so on. Similarly, I think we should follow this design.

In that case, hash of composite PK can be implemented by `hash(col1) * 31 + hash(col2)` or something similar. (Better to do a little investigation on PG/MySQL/..)

Reply:
If I understand correctly. If we can find a stable enough way to distribute the key to the vnode evenly, then we can remove the crc32 from the vnode calculation process.

Reply:
By the way, redis uses crc16 to distribute the keys in the cluster across 16384 slots, which seems pretty good

Reply:
> If I understand correctly. If we can find a stable enough way to distribute the key to the vnode evenly, then we can remove the crc32 from the vnode calculation process.

My understanding is that the `hash()` function itself is supposed to evenly distribute data. For example, strings may use crc16 (like Redis), integer may use `mod` or somehing more complex. We are actually build a 2-layer hash, and that's the most weird part to me.

Reply:
If I understand correctly, our current hash strategy is to identify the columns that need hashing, and then perform a crc32 operation on the hash keys of each row. The resulting HashCode is directly converted into a vnode, which means we only hash once.